File: server/src/collectors/earthquakeNetworkCitizen.js
/**
 * Earthquake Network — citizen smartphone-accelerometer seismograph.
 * https://www.earthquakenetwork.it/realtime/
 *
 * Crowdsourced quake detections from a global network of phones running
 * the EQN app. Complements `wolfx-eew` / `p2pquake-jma` with grass-roots
 * detections that surface micro-events JMA may not publish (or publish later).
 */

import { intelEnvelope, intelUid } from '../utils/intelHelpers.js';
import { fetchHead } from './_liveHelpers.js';

const SOURCE_ID = 'earthquake-network-citizen';
const PROBE_URL = 'https://www.earthquakenetwork.it/realtime/';

export default async function collectEarthquakeNetworkCitizen() {
  const live = await fetchHead(PROBE_URL).catch(() => false);
  return intelEnvelope({
    sourceId: SOURCE_ID,
    items: [{
      uid: intelUid(SOURCE_ID, 'portal'),
      title: 'Earthquake Network (citizen smartphones)',
      summary: 'Crowdsourced quake detections from global smartphone accelerometers',
      link: PROBE_URL,
      language: 'en',
      published_at: new Date().toISOString(),
      tags: ['earthquake', 'citizen-science', 'eqn', live ? 'reachable' : 'unreachable'],
      properties: { operator: 'Earthquake Network', reachable: live },
    }],
    live,
    description: 'EQN — citizen smartphone-accelerometer seismograph',
  });
}

File: native/collectors/sources/earthquake_network_citizen.c
/* collectors/environment/sources/earthquake_network_citizen.c
 * Port of server/src/collectors/earthquakeNetworkCitizen.js — fetchHead
 * reachability probe → ONE intel item (uid earthquake-network-citizen|portal).
 * OTHER portal-status family (lib/probe.h). */
#include "../../source.h"
#include "../../lib/probe.h"
#include "../../third_party/cJSON.h"
#include <stdio.h>
#include <stdlib.h>

#define PROBE_URL "https://www.earthquakenetwork.it/realtime/"

static int run(const source_ctx *ctx, intel_sink *sink) {
  int live = probe_head(ctx->http, PROBE_URL);
  char now[32]; probe_iso_now(now, sizeof now);

  /* tags: ['earthquake','citizen-science','eqn', live?'reachable':'unreachable'] */
  cJSON *tags = cJSON_CreateArray();
  cJSON_AddItemToArray(tags, cJSON_CreateString("earthquake"));
  cJSON_AddItemToArray(tags, cJSON_CreateString("citizen-science"));
  cJSON_AddItemToArray(tags, cJSON_CreateString("eqn"));
  cJSON_AddItemToArray(tags, cJSON_CreateString(live ? "reachable" : "unreachable"));
  char *tj = cJSON_PrintUnformatted(tags);

  /* properties: { operator, reachable } — EXACT JS key order */
  cJSON *p = cJSON_CreateObject();
  cJSON_AddStringToObject(p, "operator", "Earthquake Network");
  cJSON_AddBoolToObject(p, "reachable", live);
  char *pj = cJSON_PrintUnformatted(p);

  intel_item it = {0};
  it.remote_key   = "portal";                 /* → uid earthquake-network-citizen|portal */
  it.title        = "Earthquake Network (citizen smartphones)";
  it.summary      = "Crowdsourced quake detections from global smartphone accelerometers";
  it.link         = PROBE_URL;
  it.published_at = now;
  it.tags_json    = tj;
  it.properties_json = pj;
  int rc = sink->emit(sink, &it);

  free(tj); free(pj);
  cJSON_Delete(tags); cJSON_Delete(p);
  fprintf(stderr, "[earthquake-network-citizen] probe reachable=%d\n", live);
  return rc >= 0 ? 0 : -1;
}

static const source_def earthquake_network_citizen_def = {
  .id = "earthquake-network-citizen", .collector = "environment",
  .name = "EQN citizen smartphones", .name_ja = "EQN 市民スマホ検知",
   .update_interval_sec = 60, .run = run };
REGISTER_SOURCE(earthquake_network_citizen_def)
